src/Optimizer/Transformer/MinifyHtml.php

```diff
@@ -0,0 +1,242 @@
+<?php
+
+namespace AmpProject\Optimizer\Transformer;
+
+use AmpProject\Dom\Document;
+use AmpProject\Dom\Element;
+use AmpProject\Optimizer\Configuration\MinifyHtmlConfiguration;
+use AmpProject\Optimizer\ErrorCollection;
+use AmpProject\Optimizer\Transformer;
+use AmpProject\Optimizer\TransformerConfiguration;
+use DOMComment;
+use DOMNode;
+use DOMText;
+
+/**
+ * Transformer that that minifies HTML.
+ *
+ * @package ampproject/amp-toolbox
+ */
+final class MinifyHtml implements Transformer
+{
+
+    /**
+     * Configuration store to use.
+     *
+     * @var TransformerConfiguration
+     */
+    private $configuration;
+
+    /**
+     * Store nodes for later deletion
+     *
+     * @var array
+     */
+    private $nodesToRemove;
+
+    /**
+     * Instantiate an OptimizeAmpBind object.
+     *
+     * @param TransformerConfiguration $configuration Configuration store to use.
+     */
+    public function __construct(TransformerConfiguration $configuration)
+    {
+        $this->configuration = $configuration;
+    }
+
+    /**
+     * Apply transformations to the provided DOM document.
+     *
+     * @param Document        $document DOM document to apply the transformations to.
+     * @param ErrorCollection $errors   Collection of errors that are collected during transformation.
+     * @return void
+     */
+    public function transform(Document $document, ErrorCollection $errors)
+    {
+        if (! $this->configuration->get(MinifyHtmlConfiguration::MINIFIY)) {
+            return;
+        }
+
+        /*
+         * Store nodes for later deletion to avoid changing the tree structure
+         * while iterating the DOM.
+         */
+        $this->nodesToRemove = [];
+
+        // Recursively walk through all nodes and minify if possible.
+        $this->minifyNode($document, $this->configuration->toArray());
```

Comment:
Why is the configuration turned into an array here and needlessly copied?

The configuration is already available as an object (uses less memory) in the form of a class property (does not need to be passed around from method to method).
```suggestion
        $this->minifyNode($document);
```

Reply:
Going through the different minify methods, I see now that this is done because the configuration is being abused for storing state across method calls while traversing nodes.

I'd rather have the configuration stay clean and only deal with configuration values, and the state be extracted and used as actual method arguments. This would be `$canCollapseWhitespace` and `$inBody`.